gh-128041: Add a terminate_workers method to ProcessPoolExecutor

csm10495 opened the pull request:
Provides a way to forcefully stop all the workers in the pool

Typically this would be used as a last effort to stop all workers if unable to shutdown / join in the expected way.


<!-- gh-issue-number: gh-128041-->
* Issue: gh-128041
<!-- /gh-issue-number -->

csm10495 committed:
gh-128041 - Add a terminate_workers method to ProcessPoolExecutor

Provides a way to forcefully stop all the workers in the pool

Typically this would be used as a last effort to stop all workers if unable to shutdown / join in the expected way

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

csm10495 committed:
Fix lint

picnixz commented on Lib/concurrent/futures/process.py:
```diff
@@ -855,3 +856,29 @@ def shutdown(self, wait=True, *, cancel_futures=False):
+            signal: The signal to send to each worker process. Defaults to
+                signal.SIGINT.
+        """
+        if self._processes:
```
- To reduce the indentation level, I suggest checking `if not self._processes: return` instead.
- You should also add some tests.

csm10495 replied:
Swapped the first one. Second one is on the way.

csm10495 replied:
Added some tests.

picnixz commented on Lib/concurrent/futures/process.py:
```diff
@@ -855,3 +856,29 @@ def shutdown(self, wait=True, *, cancel_futures=False):
 
     shutdown.__doc__ = _base.Executor.shutdown.__doc__
+
+    def terminate_workers(self, signal=signal.SIGINT):
```
SIGTERM should be preferred over SIGINT as a default termination. AFAIR, SIGINT is roughly equivalent to hit "Ctrl-C" but SIGTERM is the programatic way to ask a process to terminate. However, considering the usage of this function, I also see value in using SIGINT.

csm10495 replied:
Makes sense to me. Will swap.

csm10495 committed:
Swap to SIGTERM as the default

csm10495 committed:
Add some tests

csm10495 committed:
Update some docs

picnixz commented on Lib/test/test_concurrent_futures/test_process_pool.py:
```diff
@@ -218,6 +227,60 @@ def mock_start_new_thread(func, *args, **kwargs):
+        with futures.ProcessPoolExecutor(max_workers=1) as executor:
+            executor.submit(time.sleep, 0).result()
+
+            executor.terminate_workers(9)
```
You should use `signal.SIGKILL` in this case.

picnixz commented on Lib/test/test_concurrent_futures/test_process_pool.py:
```diff
@@ -218,6 +227,60 @@ def mock_start_new_thread(func, *args, **kwargs):
+            executor.submit(time.sleep, 0).result()
+
+            executor.terminate_workers(9)
+            mock_kill.assert_called_once_with(list(executor._processes.values())[0].pid, 9)
```
Can you split `list(executor._processes.values())[0].pid` into multiple variables instead? it makes it harder to detect which of the calls could fail in case of a test failure.

picnixz commented on Lib/test/test_concurrent_futures/test_process_pool.py:
```diff
@@ -218,6 +227,60 @@ def mock_start_new_thread(func, *args, **kwargs):
+    @unittest.mock.patch('concurrent.futures.process.os.kill')
+    def test_process_pool_executor_terminate_workers_passes_signal(self, mock_kill):
+        with futures.ProcessPoolExecutor(max_workers=1) as executor:
+            executor.submit(time.sleep, 0).result()
```
Ditto here for splitting into two.

picnixz commented on Lib/test/test_concurrent_futures/test_process_pool.py:
```diff
@@ -218,6 +227,60 @@ def mock_start_new_thread(func, *args, **kwargs):
+                executor.submit(time.sleep, 0).result()
+                raise RuntimeError("Should have raised BrokenProcessPool")
+            except BrokenProcessPool:
+                pass
```
Use `self.assertRaises(BrokenProcessPool, fut.result)`

csm10495 replied:
Thank you! I couldn't remember what the unittest version of `pytest.raises` was. Swapping over.

picnixz commented on Lib/test/test_concurrent_futures/test_process_pool.py:
```diff
@@ -218,6 +227,60 @@ def mock_start_new_thread(func, *args, **kwargs):
+    def test_process_pool_executor_terminate_workers_not_started_yet(self):
+        with futures.ProcessPoolExecutor(max_workers=1) as executor:
+            # The worker has not been started yet, terminate_workers should basically no-op
+            executor.terminate_workers()
```
If you want to check for no-ops, you can mock `os.kill`.

picnixz commented on Lib/test/test_concurrent_futures/test_process_pool.py:
```diff
@@ -218,6 +227,60 @@ def mock_start_new_thread(func, *args, **kwargs):
+        with futures.ProcessPoolExecutor(max_workers=1) as executor:
+            try:
+                executor.submit(os._exit, 1).result()
+            except BrokenProcessPool:
```
Ditto for using `assertRaises`. Whenever you `except` something, you should use `assertRaises` instead.

picnixz commented on Doc/library/concurrent.futures.rst:
```diff
@@ -415,6 +415,19 @@ to a :class:`ProcessPoolExecutor` will result in deadlock.
+      of them the given signal. If the signal is not specified, the default signal
+      :data:`signal.SIGTERM` is used.
+
+      After calling :meth:`ProcessPoolExecutor.terminate_workers`, the caller, should
```
```rst
      After calling this method the caller should
```

\+ reflow the line length to fit on 80 chars.

picnixz commented on Doc/library/concurrent.futures.rst:
```diff
@@ -415,6 +415,19 @@ to a :class:`ProcessPoolExecutor` will result in deadlock.
+      :meth:`ProcessPoolExecutor.shutdown` to ensure that all other resources
+      associated with the executor are freed.
+
+      .. versionadded:: 3.14
```
```suggestion
      .. versionadded:: next
```

picnixz commented on Doc/whatsnew/3.14.rst:
```diff
@@ -306,6 +306,9 @@ concurrent.futures
   incompatible *fork* start method you must explicitly request it by
   supplying a *mp_context* to :class:`concurrent.futures.ProcessPoolExecutor`.
   (Contributed by Gregory P.  Smith in :gh:`84559`.)
+* Add :meth:`concurrent.futures.ProcessPoolExecutor.terminate_workers` as
```
```suggestion

* Add :meth:`concurrent.futures.ProcessPoolExecutor.terminate_workers` as
```

csm10495 committed:
Fix docs

picnixz commented on Misc/NEWS.d/next/Library/2024-12-17-18-53-21.gh-issue-128041.W96kAr.rst:
```diff
@@ -0,0 +1 @@
+Add a ``terminate_workers`` method to ``ProcessPoolExecutor`` to allow a way to attempt to force kill the worker processes.
```
Duplicate the What's New entry for the NEWS entry (that's what we usually do).

picnixz reviewed (commented):
Some additional comments. It would be great to have type checks test, e.g., when you pass an invalid signal value (namely, check that `os.kill` would raise a TypeError / ValueError).

csm10495 committed:
Fix docs

csm10495 committed:
PR fixes/updates

csm10495 committed:
SIGKILL doesn't exist on windows

picnixz commented on Lib/concurrent/futures/process.py:
```diff
@@ -855,3 +856,31 @@ def shutdown(self, wait=True, *, cancel_futures=False):
+        Iterates through all of the current processes and sends the given signal if
+        the process is still alive.
+
+        After terminating workers, the pool will be in a broken state and no longer usable.
```
```suggestion
        After terminating workers, the pool will be in a broken state
        and no longer usable (for instance, new tasks should not be
        submitted).
```

picnixz commented on Lib/concurrent/futures/process.py:
```diff
@@ -855,3 +856,31 @@ def shutdown(self, wait=True, *, cancel_futures=False):
+                is_alive = proc.is_alive()
+            except ValueError:
+                # The process is already exited/closed out.
+                is_alive = False
```
```suggestion
            try:
                if not proc.is_alive():
                    continue
            except ValueError:
                # The process is already exited/closed out.
                continue
```

picnixz commented on Lib/concurrent/futures/process.py:
```diff
@@ -855,3 +856,31 @@ def shutdown(self, wait=True, *, cancel_futures=False):
+                    os.kill(pid, signal)
+                except ProcessLookupError:
+                    # The process just ended before our signal
+                    pass
```
```suggestion
             try:
                 os.kill(pid, signal)
             except ProcessLookupError:
                 # The process just ended before our signal
                 pass
```

picnixz replied:
@csm10495 sorry for this one but the indentation was wrong :') (I hate web reviews..)

picnixz commented on Lib/concurrent/futures/process.py:
```diff
@@ -855,3 +856,31 @@ def shutdown(self, wait=True, *, cancel_futures=False):
 
     shutdown.__doc__ = _base.Executor.shutdown.__doc__
+
+    def terminate_workers(self, signal=signal.SIGTERM):
```
Actually, can we perhaps name it `kill`? (namely, you kill the pool's workers with the given signal) or is `kill` already taken (possibly for something else in the future)?

csm10495 replied:
I'm not sure. I thought this was explicit. I'm open for either or.

picnixz replied:
Ok, so after some research:

- `Process.terminate` -- mimics `SIGTERM`
- `Process.kill` -- sends `SIGKILL` to the process.

Many other interfaces use `terminate` for sending SIGTERM and `kill` for sending `SIGKILL`. Now, we also have [multiprocessing.Pool.terminate](https://docs.python.org/3/library/multiprocessing.html#multiprocessing.pool.Pool.terminate), so maybe we could mimic it by just naming it `terminate(signal=...)`? It would be slightly inconsistent with the others `terminate` methods since this one would be able to send a SIGKILL as well.

So it's up to you (or up to Gregory)

csm10495 replied:
I could definitely be down for `terminate` to match multiprocessing.Pool. (Though i'd still like to be able to change the signal sent so it would be a slight api difference).  I'll wait to see what Gregory thinks to either change it or leave as is.

gpshead replied:
`mulltiprocessing.Process` used under the hood here has `.terminate()` and `.kill()` methods, rather than allowing an arbitrary signal I think we should just support those two (the signal doesn't make sense on windows - those methods do the right thing there) and call those methods on the `proc`s in your loop below.  Don't use `os.kill` or `signal` ourselves.

API wise that suggests we add two methods: `.terminate_workers()` and `.kill_workers()` rather than deciding based on an argument value and supporting arbitrary other potentially-non-terminating signals.

csm10495 replied:
I thought there may be a use case for sending alternative signals, though maybe not?

I guess having kill/terminate as two different cases probably gets folks most of the way (if not fully) there. Then if a concrete use case comes in, that requires a different signal, there can either be a signal method added or a param added somewhere.

I'll update the PR when I get a chance to reflect.

csm10495 replied:
Updated. @gpshead please check again when you can. Thanks!

picnixz reviewed (commented):
Some other comments. I'll let @gpshead take over the review for a more in-depth knowledge (I'm not enough well-versed in `concurrent.futures` since I mainly use `multiprocessing` instead).

csm10495 committed:
Update Lib/concurrent/futures/process.py

Co-authored-by: Bénédikt Tran <10796600+picnixz@users.noreply.github.com>

csm10495 committed:
Apply suggestions from code review

Co-authored-by: Bénédikt Tran <10796600+picnixz@users.noreply.github.com>

csm10495 committed:
Fix indenting from suggestions

csm10495 committed:
Internally call shutdown to prevent a resource leak when calling terminate_workers

csm10495 committed:
Change test to not validate calling of os.kill since shutdown may call that

csm10495 committed:
Commit to retrigger CI

csm10495 committed:
Merge branch 'python:main' into terminate_workers

csm10495 committed:
PR feedback. Split terminate_workers into terminate_workers and kill_workers.

Remove the ability to give an arbitrary signal, though leave plumbing to make it easy enough to re-add if needed.

csm10495 committed:
Remove un-needed imports

csm10495 committed:
lint

csm10495 committed:
Harden a test a bit to ensure the correct type of kill/terminate was used

csm10495 committed:
rekick ci

gpshead commented on Lib/concurrent/futures/process.py:
```diff
@@ -627,6 +627,18 @@ class BrokenProcessPool(_base.BrokenExecutor):
     while a future was in the running state.
     """
 
+class _TerminateOrKillOperation(enum.Enum):
```
use of the enum module and an enum for this may be overkill for this internal API of the stdlib... but i'm not going to block the PR on that - if desired for import time performance reasons (enum is frowned upon due it's overblown code complexity) it can be refactored to just a couple of constant values later.

csm10495 replied:
Its sort of unfortunate to have to not use other parts of the stdlib because of import timings, but I understand. I'll leave this for now as is unless someone has a performance oriented reason to swap to constants/a dict.

picnixz replied:
1. Maybe use an IntEnum instead of an enum.Enum?
2. I'm not very fond of the use of the enumeration here as well but I won block this one. Wuld it be possible to already know how long the import now takes with -X importtime?

csm10495 committed:
Allow more time for queue to get data back

Technically this should probably be faster than the 1 second, but sometimes can be a bit slower

gpshead commented on Lib/test/test_concurrent_futures/test_process_pool.py:
```diff
@@ -228,59 +228,77 @@ def mock_start_new_thread(func, *args, **kwargs):
-        with futures.ProcessPoolExecutor(max_workers=1) as executor:
-            executor.submit(time.sleep, 0).result()
+    def test_process_pool_executor_terminate_kill_workers_not_started_yet(self):
+        for function_name in ('terminate_workers', 'kill_workers'):
```
in all of these tests, using `with self.subTest(function_name):` would make for nicer test results.  or refactoring these into _test_blah(self, function_name): helper functions that top level test_blah_terminate and test_blah_kill call so they're distinct actual test cases.

csm10495 replied:
Moments like this wish we had pytest to do `pytest.mark.parametrize`. 

I'll swap over to using subTest in the next iteration. Thanks!

gpshead reviewed (approved):
overall good, I think i'm nitpicking on minor improvements now.

csm10495 committed:
Use subTest to break up tests

csm10495 committed:
Merge branch 'main' into terminate_workers

picnixz commented on Doc/whatsnew/3.14.rst:
```diff
@@ -445,6 +445,11 @@ contextvars
+  :meth:`concurrent.futures.ProcessPoolExecutor.kill_workers` as
+  ways to terminate or kill all living worker processes in the given pool.
+  (Contributed by Charles Machalow in :gh:`128043`.)
+
```
```suggestion
 
* Add :meth:`concurrent.futures.ProcessPoolExecutor.terminate_workers` and
  :meth:`concurrent.futures.ProcessPoolExecutor.kill_workers` as
  ways to terminate or kill all living worker processes in the given pool.
  (Contributed by Charles Machalow in :gh:`128043`.)


```

picnixz commented on Lib/concurrent/futures/process.py:
```diff
@@ -855,3 +868,65 @@ def shutdown(self, wait=True, *, cancel_futures=False):
 
     shutdown.__doc__ = _base.Executor.shutdown.__doc__
+
+    def _terminate_or_kill_workers(self, operation: _TerminateOrKillOperation):
```
I don't think we add typing in general and we leave it to typeshed. Otherwise I would add a return annotation as well.

picnixz commented on Lib/concurrent/futures/process.py:
```diff
@@ -855,3 +868,65 @@ def shutdown(self, wait=True, *, cancel_futures=False):
+        and no longer usable (for instance, new tasks should not be
+        submitted).
+        """
+        if operation not in _TerminateOrKillOperation._member_map_.values():
```
```suggestion
        if operation not in _TerminateOrKillOperation:
```

I rhink it's sufficient but please check.

picnixz commented on Lib/concurrent/futures/process.py:
```diff
@@ -855,3 +868,65 @@ def shutdown(self, wait=True, *, cancel_futures=False):
+                continue
+
+            try:
+                if operation == _TerminateOrKillOperation.TERMINATE:
```
You may use a match statement instead. Hopefully it should work whether you use the enumeration nember or its value (again check it)

picnixz commented on Lib/concurrent/futures/process.py:
```diff
@@ -855,3 +868,65 @@ def shutdown(self, wait=True, *, cancel_futures=False):
+        and no longer usable (for instance, new tasks should not be
+        submitted).
+        """
+        return self._terminate_or_kill_workers(operation=_TerminateOrKillOperation.TERMINATE)
```
Can we have it under 80 chars? I think it's larger than that.

csm10495 replied:
I generally tried to do this.. some places (esp in the tests) it got a bit tough though.

picnixz commented on Lib/test/test_concurrent_futures/test_process_pool.py:
```diff
@@ -218,6 +228,81 @@ def mock_start_new_thread(func, *args, **kwargs):
 
+    def test_process_pool_executor_terminate_kill_workers(self):
+        for function_name in ('terminate_workers', 'kill_workers'):
+            with self.subTest(function_name=function_name):
```
```suggestion
            with self.subTest(function_name):
```

And ditto for the other places

picnixz commented on Lib/test/test_concurrent_futures/test_process_pool.py:
```diff
@@ -218,6 +228,81 @@ def mock_start_new_thread(func, *args, **kwargs):
+                    getattr(executor, function_name)()
+                    worker_process.join()
+
+                    if function_name == 'terminate_workers' or sys.platform == 'win32':
```
Instead of a loop you could refactor that test into a method which takes the function name and the expected exit code:

```py
def tesr_terminate(self):
    exit_code = ...
    self.do_test("terminate_workers", exit_code)
```

The other tests can also be exempt of a loop to reduce the indentation level but since they are already generic it's fine to keep them as is.

csm10495 replied:
I did something similar, yet also different since i found an existing parameterize fixture in the test suite already.

picnixz commented on Lib/test/test_concurrent_futures/test_process_pool.py:
```diff
@@ -218,6 +228,81 @@ def mock_start_new_thread(func, *args, **kwargs):
+            executor._terminate_or_kill_workers = unittest.mock.Mock()
+            executor.kill_workers()
+
+        executor._terminate_or_kill_workers.assert_called_once_with(operation=futures.process._TerminateOrKillOperation.KILL)
```
Can we have it under 80 chars? I would advise having something like:

```py
operation= ...
executor._terminate...
```

picnixz reviewed (commented):
There are a few style issues I want to address and some refactoring possibilities. I'm on mobile so it's a bit hard to review and I'm leaving for a few days. So, if possible, I'd like to hold the merge until Sunday just to address those last points. TiA

csm10495 committed:
Apply suggestions from code review

Co-authored-by: Bénédikt Tran <10796600+picnixz@users.noreply.github.com>

csm10495 committed:
PR feedback: swap to dict with constants, better subtest parameterization, lint

csm10495 committed:
swap to using context in the test

csm10495 committed:
trailing whitespace

picnixz commented on Doc/library/concurrent.futures.rst:
```diff
@@ -415,6 +415,30 @@ to a :class:`ProcessPoolExecutor` will result in deadlock.
+   .. method:: kill_workers()
+
+      Attempt to kill all living worker processes immediately by calling
+      :meth:`Process.terminate <multiprocessing.Process.kill>` on each of them.
```
```suggestion
      :meth:`Process.kill <multiprocessing.Process.kill>` on each of them.
```

picnixz commented on Lib/test/test_concurrent_futures/test_process_pool.py:
```diff
@@ -1,13 +1,17 @@
 
 from test import support
 from test.support import hashlib_helper
+from test.test_importlib.metadata.fixtures import parameterize
```
Wait, we had that kind of trick in the stdlib?????

csm10495 replied:
Sort of kind of haha

picnixz replied:
By the way we can probably move it to a new `test.support.fixtures` in a follow-up PR if you want.

picnixz commented on Lib/test/test_concurrent_futures/test_process_pool.py:
```diff
@@ -218,6 +233,81 @@ def mock_start_new_thread(func, *args, **kwargs):
+    @parameterize(*TERMINATE_OR_KILL_PARAMS)
+    def test_process_pool_executor_terminate_kill_workers_stops_pool(self, function_name):
+        with futures.ProcessPoolExecutor(max_workers=1) as executor:
+            executor.submit(time.sleep, 0).result()
```
```suggestion
            task = executor.submit(time.sleep, 0)
            task.result()
```

I prefer having the calls split in case of an error. Not sure however if `task.result()` should return None.

picnixz commented on Misc/NEWS.d/next/Library/2024-12-17-18-53-21.gh-issue-128041.W96kAr.rst:
```diff
@@ -0,0 +1,4 @@
+* Add :meth:`concurrent.futures.ProcessPoolExecutor.terminate_workers` and
+  :meth:`concurrent.futures.ProcessPoolExecutor.terminate_workers` as
+  ways to terminate or kill all living worker processes in the given pool.
+  (Contributed by Charles Machalow in :gh:`128043`.)
```
```suggestion
Add :meth:`concurrent.futures.ProcessPoolExecutor.terminate_workers` and
:meth:`concurrent.futures.ProcessPoolExecutor.terminate_workers` as
ways to terminate or kill all living worker processes in the given pool.
(Contributed by Charles Machalow in :gh:`128043`.)
```

picnixz commented on Lib/test/test_concurrent_futures/test_process_pool.py:
```diff
@@ -218,6 +233,81 @@ def mock_start_new_thread(func, *args, **kwargs):
+            elif function_name == 'kill_workers':
+                self.assertEqual(worker_process.exitcode, -signal.SIGKILL)
+
+            self.assertRaises(queue.Empty, q.get, timeout=1)
```
```py
    def _test_process_pool_executor_check_signal(self, meth_name, signal):
        manager = self.get_context().Manager()
        q = manager.Queue()

        with futures.ProcessPoolExecutor(max_workers=1) as executor:
            executor.submit(_put_sleep_put, q)

            # We should get started, but not finished since we'll terminate the
            # workers just after
            self.assertEqual(q.get(timeout=5), 'started')

            worker_process = list(executor._processes.values())[0]
            getattr(executor, meth_name)()
            worker_process.join()
            self.assertEqual(worker_process.exitcode, signal)
            self.assertRaises(queue.Empty, q.get, timeout=1)

    def test_process_pool_executor_terminate_workers_signal(self):
        signal = -signal.SIGTERM
        self._test_process_pool_executor_check_signal("terminate_workers", signal)

    def test_process_pool_executor_kill_workers_signal(self):
        if sys.platform == "win32":
            # On windows, kill and terminate both send SIGTERM
            signal = -signal.SIGTERM
        else:
            signal = -signal.SIGKILL
        self._test_process_pool_executor_check_signal("kill_workers", signal)
```

I would prefer something like this. Please, check that the code is correctly indented and doesn't have tabs instead of spaces, and doesn't have trailing spaces. TiA

picnixz commented on Lib/test/test_concurrent_futures/test_process_pool.py:
```diff
@@ -218,6 +233,81 @@ def mock_start_new_thread(func, *args, **kwargs):
+            executor._terminate_or_kill_workers = unittest.mock.Mock()
+            executor.kill_workers()
+
+        executor._terminate_or_kill_workers.assert_called_once_with(operation=futures.process._KILL)
```
```py
    def _test_process_pool_executor_stop_workers(self, meth_name, operation):
        mock_fn = unittest.mock.Mock()
        with futures.ProcessPoolExecutor(max_workers=1) as executor:
            executor._terminate_or_kill_workers = mock_fn
            getattr(executor, meth_name)()
        mock_fn.assert_called_once_with(operation=operation)

    def test_process_pool_executor_terminate_workers(self):
        meth, op = TERMINATE_WORKERS_METH, futures.process._TERMINATE
        self._test_process_pool_executor_stop_workers(method, op)

    def test_process_pool_executor_kill_workers(self):
        meth, op = KILL_WORKERS_METH, futures.process._KILL
        self._test_process_pool_executor_stop_workers(method, op)
```

And somewhere else, define `TERMINATE_WORKERS_METH = "terminate_workers"` and `KILL_WORKERS_METH = "kill_workers"`, either directly or by using `cls.terminate_workers.__name__` so that we pick up the refactoring if needs arise. If you think it's too much of a hassle, just hardcode them.

picnixz commented on Lib/concurrent/futures/process.py:
```diff
@@ -855,3 +863,66 @@ def shutdown(self, wait=True, *, cancel_futures=False):
+        submitted).
+        """
+        if operation not in _TERMINATE_OR_KILL_OPERATION:
+            raise ValueError(f"Unsupported operation: {operation}")
```
```suggestion
            raise ValueError(f"Unsupported operation: {operation!r}")
```

picnixz commented on Lib/concurrent/futures/process.py:
```diff
@@ -855,3 +863,66 @@ def shutdown(self, wait=True, *, cancel_futures=False):
 
     shutdown.__doc__ = _base.Executor.shutdown.__doc__
+
+    def _terminate_or_kill_workers(self, operation):
```
```suggestion
    def _force_shutdown(self, operation):
```

How about this name? because we could say that we can generalize it later with other callback operation.

csm10495 replied:
Perfect. I had a lot of trouble coming up with a better sound name. Thanks!

picnixz commented on Lib/concurrent/futures/process.py:
```diff
@@ -626,6 +626,14 @@ class BrokenProcessPool(_base.BrokenExecutor):
+_TERMINATE_OR_KILL_OPERATION = {
+    _TERMINATE,
+    _KILL
+}
```
```suggestion
_SHUTDOWN_CALLBACK_OPERATION = {_TERMINATE, _KILL}
```

picnixz reviewed (commented):
While I think the parametrize is great, I think it's still nicer to have separate methods. Now, I suggest not applying any suggestion using the web UI but rather make separate commits depending on what is changed.

I think we can rename `_terminate_or_kill_workers` into `_force_shutdown()` as we may or may not extend the functionality later. What is however sure is that we'll anyway shutdown the processes.

Also, I don't think you need to prefix every test method by "test_process_pool_executor". If the test is for a futures.ProcessPoolExecutor only, then just using `test_executor_*` is sufficient (we're already in the `ProcessPoolExecutorTest` class).

By the way, should always test `futures.ProcessPoolExecutor` or can we use some `self.executor_type()` instead?

csm10495 committed:
Various pr feedbacks

Drop extra prefix on tests, fix formating on gh issue file

Separate a couple operations into different lines

csm10495 committed:
PR feedback: swap name of terminate_or_kill to force_shutdown

csm10495 committed:
PR feedback: swap test names

csm10495 committed:
PR feedback: use self.executor_type instead of ProcessPoolExecutor directly

csm10495 committed:
Add constants for terminate/kill methods

picnixz commented on Misc/NEWS.d/next/Library/2024-12-17-18-53-21.gh-issue-128041.W96kAr.rst:
```diff
@@ -0,0 +1,4 @@
+Add :meth:`concurrent.futures.ProcessPoolExecutor.terminate_workers` and
+:meth:`concurrent.futures.ProcessPoolExecutor.terminate_workers` as
```
```suggestion
:meth:`concurrent.futures.ProcessPoolExecutor.kill_workers` as
```

picnixz commented on Lib/test/test_concurrent_futures/test_process_pool.py:
```diff
@@ -218,6 +235,84 @@ def mock_start_new_thread(func, *args, **kwargs):
+            executor._force_shutdown = unittest.mock.Mock()
+            executor.kill_workers()
+
+        executor._force_shutdown.assert_called_once_with(operation=futures.process._KILL)
```
```suggestion
        mock_fn = unittest.mock.Mock()
        with self.executor_type(max_workers=1) as executor:
            executor._force_shutdown = mock_fn
            executor.kill_workers()

        mock_fn.assert_called_once_with(operation=futures.process._KILL)
```

picnixz commented on Lib/test/test_concurrent_futures/test_process_pool.py:
```diff
@@ -218,6 +235,84 @@ def mock_start_new_thread(func, *args, **kwargs):
+            executor._force_shutdown = unittest.mock.Mock()
+            executor.terminate_workers()
+
+        executor._force_shutdown.assert_called_once_with(operation=futures.process._TERMINATE)
```
```suggestion
    def test_terminate_workers(self):
        mock_fn = unittest.mock.Mock()
        with self.executor_type(max_workers=1) as executor:
            executor._force_shutdown = mock_fn
            executor.terminate_workers()

        mock_fn.assert_called_once_with(operation=futures.process._TERMINATE)
```

picnixz reviewed (commented):
The rationale behind my suggestion is twofold:

- It helps wrapping lines under 80 chars (PEP-8)
- It makes it cleaner for debugging in case of failures.

However, I think using `parametrize` is fine now. I think what bothered me most was `_terminate_or_kill` naming because. But now it looks better. I would still want you to respect 80 chars wrap.

csm10495 committed:
feedback to get below 80 chars per pep8

csm10495 committed:
Merge branch 'main' into terminate_workers

gpshead commented on Lib/test/test_concurrent_futures/test_process_pool.py:
```diff
@@ -218,6 +235,86 @@ def mock_start_new_thread(func, *args, **kwargs):
+            else:
+                self.fail(f"Unknown operation: {function_name}")
+
+            self.assertRaises(queue.Empty, q.get, timeout=1)
```
I do suspect we may see this come up as occasionally flaky in buildbot or CI systems as it is depend on the timing of the sleeps and kills which really can't be guaranteed on a loaded system.  if so, _put_sleep_put can have its sleep increased.  lets see how it goes first.
